### ISSUE TITLE
network-manager: uses hardcoded path to ping

### ISSUE THREAD

**Issue by peterhoeg** [NixOS/nixpkgs#14103]
```
Mar 21 21:54:33 margot NetworkManager[756]: <warn>  (enp8s0): ping: could not spawn /usr/bin/ping: Failed to execute child process "/usr/bin/ping" (No such file or directory)
```

* System: 16.09pre78848.9549257 (Flounder)
* Nix version: nix-env (Nix) 1.11.2
* Nixpkgs version: 16.09pre78848.9549257

## Describe your issue here

network-manager tries to call ping via an absolute path and fails.

### Expected result

nm should be using the proper path in /nix/store

### Actual result

It tries ```/usr/bin/ping``` and fails.

### Steps to reproduce

Install network-manager and reboot.

**Comment by domenkozar** [comment]
How to trigger the ping?

**Comment by peterhoeg** [comment]
Actually, I don't know what exactly network-manager is trying to do, but I had a quick look through the nm source, and it seems like adding this to the preConfigure phase does the trick:

Refer to: https://github.com/peterhoeg/nixpkgs/blob/nm_ping/pkgs/tools/networking/network-manager/default.nix#L18-L19
```
substituteInPlace src/devices/nm-device.c --replace /usr/bin/ping /var/setuid-wrappers/ping
```
I have managed to build a new packge with this, but as I haven't figured out how to install that to the system profile for it to actually take effect, I haven't tested it yet.

**Comment by joachifm** [comment]
Won't that make the network-manager package NixOS only?

**Comment by peterhoeg** [comment]
It's not changes to upstream - we patch in `network-manager/default.nix` or am I misunderstanding your question?

**Comment by joachifm** [comment]
If you hard-code a call-out to a path that typically only ever exists on NixOS, then the resulting binary will be NixOS only (i.e., it will fail to work on !NixOS out of the box). There are several packages that do this, already, but IMO that's not a good thing.

**Comment by abbradar** [comment]
I think that we can apply a patch in this particular case at least because:

1. NetworkManager is a system service, usually provided by the distribution. It's difficult to integrate services built with Nix into another system (you would need to somehow add all needed systemd, dbus and whatever services to your system or even do things some another way if your distro is non-systemd etc etc)
2. `/usr/sbin/ping` is as arbitrary path as `/var/setuid-wrappers/ping`. Yes, it's more standard, but the main problem is that it is hardcoded in the first place. While it's like this, at least we can make it work on NixOS as our primary Linux target.

**Comment by abbradar** [comment]
@peterhoeg To apply it to the system profile, do the following (assuming that you have made your changes on top of `nixos-unstable` or your other channel of choice to minimize rebuilds): `nixos-rebuild switch -I nixpkgs=path/to/your/nixpkgs`

**Comment by peterhoeg** [comment]
@joachifm, I didn't consider NixOS vs Nix on other platforms. That makes sense. I think @abbradar makes a good point about the use of nm on !NixOS, and we can basically choose between these:

1. keep network-manager partially broken on NixOS due to the hardcoded path
2. break nm partially on !NixOS with the changed path

My guess is that this change would unbreak more setups than it would break.

And btw, the mentioned change does work now that I've had the chance to try it out. Opened #14155

**Comment by peti** [comment]
NM tries to ping the default gateway on the other side of the connection to verify that IP connectivity has been established successfully.

IMHO, the proper way to do this is to let NM call `ping` without an absolute path and trust that the daemon has been started in an environment that has `ping` in `$PATH`.

**Comment by abbradar** [comment]
That is, if it uses `exec*p` family of functions -- this should be checked. I also have another idea (copying from #14155) how to fix this properly: as what user does NM usually run? If it's root or someone with raw network packet privileges, we can just replace path to ping in Nix store because we don't need SUID bit. For me at least it runs as root.

**Comment by peterhoeg** [comment]
Fixed by: #14242
